### ISSUE TITLE
Parent job does not cancel child jobs

### ISSUE THREAD

**Issue by CodeRedDev** [Kotlin/kotlinx.coroutines#1852]
This issue is more of a question as I'm slowly beginning to really understand how coroutines work but I'm not yet sure if the way I'm using coroutines is right.

I have this class that processes requests and when done returns its results to the caller. I want to make this process cancellable but unfortunately the way I thought it would work doesn't work.

(Don't worry about some strange names. I had to censor the code)

```kotlin
class Processor(private val client: ApiClient) {
    ...
    private val coContext = newSingleThreadContext(Processor::class.java.simpleName)
    private var parentJob: Job? = null

    fun processRequests(requests: List<Request>, callback: (List<Response) -> Unit) {
        GlobalScope.launch(coContext) {
            parentJob?.cancelAndJoin()
            parentJob = launch {
                val completedRequests = mutableListOf<Response>()
                batchProcessRequests(requests, completedRequests)
                callback(completedRequests)
            }
        }
    }

    private suspend fun batchProcessRequests(requests: List<Request>, completedRequests: MutableList<Response>) = coroutineScope {
        requests.forEach { request ->
            ensureActive()
            launch(Dispatchers.IO) {
                val response = client.getResponse(request) // Makes a Retrofit call. See below
                completedRequests.add(response)
            }
        }
    }
}

class ApiClient() {
    private val api = Api.create(...)

    suspend fun getResponse(request: Request): Response {
        // Do some extra stuff before sending the request
        ...
        return api.requesting(...)
    }
}

interface Api {
    @GET("...")
    suspend fun requesting(...): Response

    ...
}
```

If I'm right the Retrofit suspend function works like `deferred.await()` that's why I'm launching another coroutine to not waste time. When I'm calling `processRequest()` shortly after a first call to it, it will call `cancelAndJoin()` and wait for the `parentJob` to complete its cancellation.  Unfortunately this cancel call will not cancel any of the inner launched coroutines that are getting the response for a request. Only after completion of `batchProcessRequests()` the `parentJob` will be cancelled successfully.

Shouldn't `parentJob` cancel all its children almost immediately?

**Comment by elizarov** [comment]
I don't see any problem that would prevent the cancellation of all the children. Can you provide a self-contained example that we can run to see the problem?

**Comment by CodeRedDev** [comment]
I just once again checked everything I used in the process of testing and I think I found the source of the problem I encountered. I was mocking the `ApiClient` with mockk to not make any calls to the real API in tests. Due to using the wrong mockk functions for answering a call to `getResponse()` I kind of turned this cooperative coroutine into a non cooperative. 

For everyone who's interested in what code caused this problem:
```kotlin
private val apiClient = mockk<ApiClient>().apply {
    coEvery { getResponse(any()) } answers  {
        Thread.sleep(1000L) // Simulate passed time
        Response()
    }
}
```
Instead this made it work as expected:

```kotlin
private val apiClient = mockk<ApiClient>().apply {
    coEvery { getResponse(any()) } coAnswers  {
        delay(1000L) // Simulate passed time
        Response()
    }
}
```
Thank you very much for your feedback Roman. It helped me to change my focus from the processing to the testing component.

I'll close this issue.
